plugins/modem-manager/fu-qmi-pdc-updater.c

```diff
@@ -0,0 +1,432 @@
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (error_code != 0) {
+		/* when a given mcfg file already exists in the device, an "invalid id" error is returned;
+		 * the error naming here is a bit off, as the same protocol error number is used both for
+		 * 'invalid id' and 'invalid qos id'
+		 */
+		if (error_code == QMI_PROTOCOL_ERROR_INVALID_QOS_ID) {
+			g_debug ("file already available in device");
+			g_main_loop_quit (ctx->mainloop);
+			return;
+		}
+
+		g_set_error (&ctx->error, G_IO_ERROR, G_IO_ERROR_FAILED,
+			     "couldn't load mcfg: %s", qmi_protocol_error_get_string ((QmiProtocolError) error_code));
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (qmi_indication_pdc_load_config_output_get_frame_reset (output, &frame_reset, NULL) && frame_reset) {
+		g_set_error (&ctx->error, G_IO_ERROR, G_IO_ERROR_FAILED,
+			     "couldn't load mcfg: sent data discarded");
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!qmi_indication_pdc_load_config_output_get_remaining_size (output, &remaining_size, &ctx->error)) {
+		g_prefix_error (&ctx->error, "couldn't load remaining size: ");
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (remaining_size == 0) {
+		g_debug ("finished loading mcfg");
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	g_debug ("loading next chunk (%u bytes remaining)", remaining_size);
+	fu_qmi_pdc_updater_load_config (ctx);
+}
+
+static void
+fu_qmi_pdc_updater_load_config_ready (GObject *qmi_client, GAsyncResult *res, gpointer user_data)
+{
+	WriteContext *ctx = (WriteContext *) user_data;
+	g_autoptr(QmiMessagePdcLoadConfigOutput) output = NULL;
+
+	output = qmi_client_pdc_load_config_finish (QMI_CLIENT_PDC (qmi_client), res, &ctx->error);
+	if (output == NULL) {
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!qmi_message_pdc_load_config_output_get_result (output, &ctx->error)) {
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	/* after receiving the response to our request, we now expect an indication
+	 * with the actual result of the operation */
+	g_assert (ctx->indication_id == 0);
+	ctx->indication_id = g_signal_connect (ctx->qmi_client, "load-config",
+					       G_CALLBACK (fu_qmi_pdc_updater_load_config_indication), ctx);
+
+	/* don't wait forever */
+	g_assert (ctx->timeout_id == 0);
+	ctx->timeout_id = g_timeout_add_seconds (5, fu_qmi_pdc_updater_load_config_timeout, ctx);
+}
+
+static void
+fu_qmi_pdc_updater_load_config (WriteContext *ctx)
+{
+	g_autoptr(QmiMessagePdcLoadConfigInput) input = NULL;
+	g_autoptr(GArray) chunk = NULL;
+	gsize full_size;
+	gsize chunk_size;
+
+	input = qmi_message_pdc_load_config_input_new ();
+	qmi_message_pdc_load_config_input_set_token (input, ctx->token++, NULL);
+
+	full_size = g_bytes_get_size (ctx->blob);
+	chunk_size = (((ctx->offset + QMI_LOAD_CHUNK_SIZE) > full_size) ?
+		      (full_size - ctx->offset) :
+		      QMI_LOAD_CHUNK_SIZE);
```

Comment:
I think an `if` would make this a lot clearer.

Reply:
ok